glean/src/server_events.rs

```diff
@@ -0,0 +1,309 @@
+        // times are ISO-8601 strings, e.g. "2023-12-19T22:09:17.440Z"
+        let now = Utc::now().to_rfc3339();
+        PingInfo {
+            seq: 0,
+            start_time: now.clone(),
+            end_time: now,
+        }
+    }
+
+    fn create_ping(
+        &self,
+        document_type: &str,
+        config: &RequestInfo,
+        payload: &PingPayload,
+    ) -> Ping {
+        let payload_json =
+            serde_json::to_string(payload).expect("unable to marshal payload to json.");
+        let document_id = Uuid::new_v4().to_string();
+        Ping {
+            document_namespace: self.app_id.clone(),
+            document_type: document_type.to_string(),
+            document_version: "1".to_string(),
+            document_id,
+            user_agent: Some(config.user_agent.clone()),
+            ip_address: Some(config.ip_address.clone()),
+            payload: payload_json,
+        }
+    }
+
+    /// Method called by each ping-specific record method.
+    /// The goal is to construct the ping, wrap it in the envelope
+    /// and print to stdout.
+    fn record(
+        &self,
+        document_type: &str,
+        request_info: &RequestInfo,
+        metrics: Metrics,
+        events: Vec<GleanEvent>,
+    ) {
+        let telemetry_payload: PingPayload = PingPayload {
+            client_info: self.create_client_info(),
+            ping_info: GleanEventsLogger::create_ping_info(),
+            metrics,
+            events,
+        };
+
+        let ping: Ping = self.create_ping(document_type, request_info, &telemetry_payload);
+
+        let envelope: LogEnvelope = LogEnvelope {
+            log_type: GLEAN_EVENT_MOZLOG_TYPE.to_string(),
+            fields: ping,
+        };
+        let envelope_json =
+            serde_json::to_string(&envelope).expect("unable to marshal payload to json.");
+        println!("{}", envelope_json);
+    }
+}
+
+// Code below is generated based on the provided `metrics.yaml` file:
+
+// Metrics of the `event` type. Anything defined in `extra_keys` has it's own struct field.
+// The appended `Event` term to any metric of the event type implies the ping event.
+
+pub struct SyncstorageGetCollectionsEvent {
+    // metadata for event in `extra_keys`
+}
+
+// Implementing the EventsPingEvent trait for the generated struct SyncstorageGetCollectionsEvent
+impl EventsPingEvent for SyncstorageGetCollectionsEvent {
+    /// Create a GleanEvent for the above-defined Event struct (SyncstorageGetCollectionsEvent).
+    /// Any metadata `extra` values are passed into the extra HashMap.
+    fn glean_event(&self) -> GleanEvent {
+        // Any `extra_keys` will be output below to be inserted into `extra`.
+        // If there are none, an empty, immutable HashMap is created.
+        let extra: HashMap<String, String> = HashMap::new();
+
+        new_glean_event("syncstorage", "get_collections", extra)
+    }
+}
+
+/// Marker trait for events per ping
+pub trait EventsPingEvent {
```

Comment:
I'm not sure we need this trait definition. 
All it does is generate a GleanEvent, which could be done using a Default constructor.

Reply:
This is auto generated for the many possible `metric.yaml` defined `Events` (which in this case really refers to pings, since events plural connotes not the metric type of event, but the ping event). This is to say that the marker trait is generated with the `glean_parser` code which we want to leave in place, just for conformity if that's ok